### PR TITLE
feat: added section for multipart/form-data and binary

### PR DESCRIPTION
**Pull request by spark-c**
Adds a section/heading on the [Capturing Form Data](https://docs.appsmith.com/core-concepts/data-access-and-binding/capturing-data-write/capture-form-data) page to give instructions on configuring and using the **MULTIPART_FORM_DATA** query encoding. This page already had instructions for submitting a couple of other types/formats of form data to APIs, so it seemed like a good home.

[First draft on Notion](https://www.notion.so/appsmith/Multipart-and-Binary-data-be27baa167c546d5a92abbff8e6f15b5)
[Resolves appsmithorg/appsmith#12415](https://github.com/appsmithorg/appsmith/issues/12415)

### CODE REVIEW

**Comment by jnikhila** [comment]
> @jnikhila This should be somewhere in our REST API docs, correct? somewhere in https://docs.appsmith.com/core-concepts/connecting-to-data-sources/authentication/connect-to-apis
> and the info can also be cross-referenced in the Filepicker widget.

Yes, this should be the part of configure API.

**Comment by jnikhila** [comment]
> > @jnikhila This should be somewhere in our REST API docs, correct? somewhere in https://docs.appsmith.com/core-concepts/connecting-to-data-sources/authentication/connect-to-apis
> > and the info can also be cross-referenced in the Filepicker widget.
> 
> Yes, this should be part of the `configure API`.

@spark-c - could you please co-ordinate with @harshilp24 and make the changes.

**Comment by spark-c** [comment]
Sure thing @jnikhila

**Comment by spark-c** [comment]
@jnikhila @harshilp24 () Looking at the [connect-to-apis](https://appsmith-docs-l36ocl7qw-get-appsmith.vercel.app/core-concepts/connecting-to-data-sources/authentication/connect-to-apis) page where we're moving the new multipart form data sections, it seems like the [**Body**](https://appsmith-docs-l36ocl7qw-get-appsmith.vercel.app/core-concepts/connecting-to-data-sources/authentication/connect-to-apis#body) sub-heading would be the best fit.

However, it looks like this page is meant to be an overview of the REST API interface / introduction to the different parts of a request. It might seem a bit odd/not match the tone of the page if the content in the **Body** subheading suddenly gets really specific about multipart and raw binary form data. For easy reference, I'll attach the new sections to the bottom of this comment as they are currently.

Do you have advice for fitting this into the page? I'm not sure of the best course for it.

(N.B. This is just a copy/paste, so links and imgs are broken below.)
> ## Submitting multipart and binary data
> 
> Appsmith supports sending file data in API queries via `multipart/form-data` encoding, which is commonly expected by API endpoints that handle file uploads. Multipart requests can include several different types of data within them, such as a file along with some other related metadata.
> 
> To set up your query to send with `multipart/form-data` encoding, navigate to its query editor screen, click the **Body** tab, and find the **MULTIPART_FORM_DATA** tab beneath it.
> 
> Use multipart form encoding in your API request
> 
> To submit a file as a multipart input, use a [Filepicker widget](/reference/widgets/filepicker) to upload a file to your application. Once it has been uploaded, you can bind `{{ FilePicker1.files[0] }}` as a value in your API's multipart request body. Be sure to select "File" in the datatype dropdown. If you would like to submit multiple files in the same request key, you can alternatively use `{{ FilePicker1.files }}` to include the entire contents of the Filepicker widget.
> 
> The file data from the Filepicker widget can be in any data format (base64 or binary) according to the requirements of the endpoint that you're connecting to.
> 
> You can also pass plain text values in your multipart request by selecting the "Text" option in the datatype dropdown. If you wish to pass multiple plain text values under the same key, be sure to use the "Array" option instead.
> 
> Use "File," "Array," and "Text" data types in your multipart-encoded request.
> 
> ```javascript
> // type: Text
> {{ Text1.text }}
> ```
> 
> ```javascript
> // type: Array
> {{[ Text1.text, Text2.text, "hello, world"]}}
> ```
> 
> ### Raw body binary
> 
> If your endpoint can't accept multipart-encoded data and requires raw body binary instead, choose the **RAW** tab under the query **Body** tab instead of MULTIPART_FORM_DATA. In this case, you would pass the `data` property of your file to the query instead of the file object itself, because the endpoint expects only raw binary data:
> 
> ```javascript
> // Binary data in the RAW format
> {{ Filepicker1.files[0]?.data }}
> ```
> 
> :::note
> The previous example uses [Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) via the JavaScript `?` operator. This way, if the `files[0]` doesn't exist, attempting to access `data` returns `undefined` instead of raising an error.
> :::
> 
> Be sure to turn off the **JSON Smart Substitution** setting for this query in the query settings. This option is normally useful for helping to cast data into the correct JSON formats, however it can be problematic when used with RAW body binary encoding.

**Comment by jnikhila** [comment]
> @spark-c - Agreed. The configuration specific to these are moved to Query settings page. @Pranay105 is working on that.

**Comment by danciaclara** [comment]
> > @spark-c - Agreed. Per my discussion with @danciaclara - we might come up with another route to follow. @danciaclara - Please share your thoughts here.

So I'm thinking let's keep this as a doc under How to Guides and then cross-reference it in Configure API's Body tab section/  Further Reading section and in FilePicker widget's Further reading section

File: website/docs/learning-and-resources/how-to-guides/README.md [hunk 1]
```diff
@@ -8,15 +8,16 @@ This document presumes you are familiar with the [core concepts](core-concepts/c
 
 ## Guide Reference
 
-1. [OAuth2 authorization for Google Sheets](oauth2-authorization-for-google-sheets.md)
```
Comment (jnikhila) [inline website/docs/learning-and-resources/how-to-guides/README.md:11]: @danciaclara - Should we have cards for these too? We can do a separate PR for this.

File: website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+---
+description: Guide for submitting multipart/form-data and binary file data via Appsmith
+sidebar_position: 17
+---
+
+# Submitting Multipart and Binary form Data
```
Comment (danciaclara) [inline website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md:6]:
```suggestion
# Send Body Data with API Requests```

File: website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+---
+description: Guide for submitting multipart/form-data and binary file data via Appsmith
```
Comment (danciaclara) [inline website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md:2]:
```suggestion
description: Guide for submitting body data with API requests on Appsmith
```

File: website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+sidebar_position: 17
+---
+
+# Submitting Multipart and Binary form Data
+
+Appsmith supports sending file data in API queries via `multipart/form-data` encoding, which is commonly expected by API endpoints that handle file uploads. Multipart requests can include several different types of data within them, such as a file along with some other related metadata. The Appsmith REST API interface can also handle submitting raw binary file data.
+
+## Multipart form data
```
Comment (danciaclara) [inline website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md:10]:
```suggestion
## Multipart/form-data
```

File: website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+// type: Text
+{{ Text1.text }}
+```
+
+```javascript
+// type: Array
+{{[ Text1.text, Text2.text, "hello, world"]}}
+```
+
+## Raw body binary
```
Comment (danciaclara) [inline website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md:34]:
```suggestion
## Raw data
```

File: website/docs/reference/widgets/filepicker.md [hunk 6]
```diff
@@ -130,15 +125,16 @@ Style properties allow you to change the look and feel of the widget.
...
+### **Further reading**
 
-The following resources will come in handy as you need to learn new tricks:
+The following resources should come in handy as you need to learn new tricks:
 
 * [Connecting to Datasources](/core-concepts/connecting-to-data-sources)
 * [Appsmith Framework](../appsmith-framework/)
 * [JavaScript Editor](../../core-concepts/writing-code/javascript-editor-beta/)
 * [Uploading a File to S3](../../learning-and-resources/how-to-guides/how-to-upload-to-s3.md)
+* [How to submit multipart form data and binary file data via API](/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data)
```
Comment (danciaclara) [inline website/docs/reference/widgets/filepicker.md:136]:
```suggestion
* [Upload File Data with API Requests](/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data)
```

File: website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+---
+description: Guide for submitting multipart/form-data and binary file data via Appsmith
+sidebar_position: 17
+---
+
+# Submitting Multipart and Binary form Data
+
+Appsmith supports sending file data in API queries via `multipart/form-data` encoding, which is commonly expected by API endpoints that handle file uploads. Multipart requests can include several different types of data within them, such as a file along with some other related metadata. The Appsmith REST API interface can also handle submitting raw binary file data.
```
Comment (danciaclara) [inline website/docs/learning-and-resources/how-to-guides/how-to-submit-multipart-and-binary-form-data.md:8]: Put this under the Multipart form data section. Here write a general intro about send body data with API requests.

**Review by danciaclara — CHANGES_REQUESTED** [review]
Let's also move this [section](https://docs.appsmith.com/core-concepts/data-access-and-binding/capturing-data-write/capture-form-data#posting-url-encoded-form-data) to this page and title the section - **URL-encoded form data**

File: website/docs/core-concepts/data-access-and-binding/capturing-data-write/capture-form-data.md [hunk 8]
```diff
@@ -33,6 +40,49 @@ Once the API / Query updates the data, the widgets on the screen need to be upda
 The Property Pane UI supports a single onSuccess Callback but multiple callbacks and conditions can be configured by clicking the JS button next to the property. Learn more about [Creating Workflows](../../writing-code/workflows.md)
 :::
 
-## Posting URL Encoded Form Data
+## Send body data with API requests
```
Comment (danciaclara) [inline website/docs/core-concepts/data-access-and-binding/capturing-data-write/capture-form-data.md:43]: @spark-c I think there is some confusion. This section will be a How to guide which is why all of review comments were made on that page itself. The was to move URL-encoded form data to that page. I think because of the confusion everything is moved here. Can you quickly fix that. This page talks about Capturing data entered in the UI.  We can cross-reference the information here but the data should reside in How to guides

**Review by danciaclara — CHANGES_REQUESTED** [review]

**Comment by danciaclara** [comment]
@spark-c can you pull the latest changes on this branch. I made some rearrangements. Removed the How to guide section and this info resides now in the REST APIs page. We can think of a better placement if necessary.

**Review by danciaclara — APPROVED** [review]
